Mempool space

Beerosagos opened the pull request:
Bitcoin Core fee estimation is not very accurate. With this change, meempool.space api will be used as a preferred estimation method per BTC fees. The fee estimations are fetched through a Shift mirror server.

Beerosagos commented on backend/coins/btc/account.go:
```diff
@@ -139,13 +148,16 @@ func NewAccount(
-			{blocks: 12, code: accounts.FeeTargetCodeLow},
-			{blocks: 6, code: accounts.FeeTargetCodeNormal},
-			{blocks: 2, code: accounts.FeeTargetCodeHigh},
+			{blocks: 6, code: accounts.FeeTargetCodeLow},
```
We'll need to update `send.feeTarget.description.*` frontend translations.

benma reviewed (commented):
I noticed another issue:

The current (old) estimator is using Bitcoin Core's `estimatesmartfee` call, which only takes into consideration past blocks. So fees are updated only when a new block is found. This is why in the app, the fees are updated also when a new block is found.

mempool.space fees take the mempool into consideration, so updating only when a block is found migt not be not sufficient to get accurate fees.

In general it was maybe not a good idea to download the fee estimates even if the user is not going to make a transaction, just viewing the account.

I propose we switch the logic to download the fees when the user needs it, i.e. when they go to the send screen and the fee targets are to be loaded, i.e. at the same time (or part of) the `account/${code}/fee-targets` API call. This would work for both the old and the new fee estimators.

Beerosagos commented:
rebased ^

Beerosagos commented on frontends/web/src/locales/en/app.json:
```diff
@@ -1560,6 +1560,10 @@
         "low": "2 hours (12 blocks)",
         "low_eth": "5 minutes or less",
         "low_ltc": "30 minutes (12 blocks)",
+        "mEconomy": "-",
```
mempool.space doesn't give a time estimation for the economy fee. It this ok?

Beerosagos commented:
Pushed a new update:
- removed the fee caching 
- created different fee targets for mempool.space
- removed the deletion of duplicated fees from `account.FeeTargets()`
- addressed the other comments

@benma PTAL :pray: 

I feel like we could refactor a bit more, moving the fee handling part from the account to the coin. But I think it's better to do it in a separate PR, to keep things simple.

benma commented on backend/accounts/feetarget.go:
```diff
@@ -58,10 +59,51 @@ const (
+}
+
+// GetFeeAmount returns the btcutil.Amount of the fee for the given FeeTargetCode.
+func (fees MempoolSpaceFees) GetFeeAmount(code FeeTargetCode) btcutil.Amount {
```
Better name could be `GetFeeRate()`, as this does not return the fee, but the fee rate (sats/vB).

Beerosagos replied:
Ack, I wasn't really satisfied about the name either

benma commented on backend/coins/btc/account.go:
```diff
@@ -476,26 +475,72 @@ func (account *Account) Notifier() accounts.Notifier {
+		_, err := util.APIGet(account.httpClient, mempoolSpaceMirror, "", 1000, mempoolFees)
+		if err != nil {
+			mempoolFees = nil
+			account.log.WithError(err).Error("Fecthing fees from " + mempoolSpaceMirror + " failed")
```
fyi there is also `Errorf`: `.Errorf("Fetching fees from %s failed", mempoolSpaceMirror)`

also: type in `Fetching`

benma commented on backend/coins/btc/account.go:
```diff
@@ -476,26 +475,72 @@ func (account *Account) Notifier() accounts.Notifier {
+			// If even that one is not available, we just offer the min relay fee.
+			if account.coin.Code() == coin.CodeBTC {
 				account.log.WithField("fee-target", feeTarget.blocks).
-					Warning("Fee could not be estimated. Taking the minimum relay fee instead")
+					Debug("Fee server not available. Trying to estimate.")
```
Isn't this log redundant with the one above (`Fecthing fees from " + mempoolSpaceMirror + " failed")`)?

Beerosagos replied:
Removed :+1:

benma commented on backend/coins/btc/account.go:
```diff
@@ -476,26 +475,72 @@ func (account *Account) Notifier() accounts.Notifier {
+	for _, feeTarget := range feeTargets {
+		var feeRatePerKb btcutil.Amount
+
+		if mempoolFees != nil {
```
this condition is true always even for TBTC or LTC as you initialize it to ` mempoolFees := &accounts.MempoolSpaceFees{}`.

Beerosagos replied:
Shoot! I don't know how I missed this during the testing. Great catch :pray:

benma commented on backend/coins/btc/account.go:
```diff
@@ -476,26 +475,72 @@ func (account *Account) Notifier() accounts.Notifier {
+// unavailable.
+func (account *Account) feeTargets() []*FeeTarget {
+	// for mainnet BTC we fetch mempool.space fees, as they should be more reliable.
+	mempoolFees := &accounts.MempoolSpaceFees{}
```
init to `nil`, set to nonnil inside the condition below - see comment below about the bug this causes otherwise.

benma reviewed (changes requested):
I run into this error:

https://github.com/BitBoxSwiss/bitbox-wallet-app/blob/4ae57210793e7cf6eb25d7a2213e51e6f05292cc/backend/accounts/feetarget.go#L43

Because the mempool space codes are not covered.

Beerosagos commented:
rebased

Beerosagos commented:
@benma Thanks for the review and for catching the bugs. I think I only checked that the fees were displayed properly in the interface and I didn't test an actual signing, so I missed the error you got. 

Now it should be good.

benma commented:
> @benma Thanks for the review and for catching the bugs. I think I only checked that the fees were displayed properly in the interface and I didn't test an actual signing, 

I also didn't do any signing - the fees dropdown is disabled until you add an address/amount, and at that point it already failed.

benma reviewed (approved):
Thanks.

When I go to the send-screen, the fee-targets endpoint is called many times, which means there are many requests to the server. I fixed this in https://github.com/BitBoxSwiss/bitbox-wallet-app/pull/2730.

tACK

Beerosagos commented:
> I also didn't do any signing - the fees dropdown is disabled until you add an address/amount, and at that point it already failed.

I was curious and checked: it actually worked with just the amount until you add an address, because it needs both the address and the amount to make the TX proposal (which was the failing part). This is how I didn't notice the error :D

> When I go to the send-screen, the fee-targets endpoint is called many times, which means there are many requests to the server. I fixed this in https://github.com/BitBoxSwiss/bitbox-wallet-app/pull/2730.

Awesome :heart:  Thanks a lot
